Add VTOrc recovery for mismatch in tablet type

GuptaManan100 opened the pull request:
<!--
  Thank you for your contribution to the Vitess project.
  How to contribute: https://vitess.io/docs/contributing/
  Please first make sure there is an open Issue to discuss the feature/fix suggested in this PR.
  If this is a new feature, please mark the Issue as "RFC".
 -->

<!-- if this PR is Work in Progress please create it as a Draft Pull Request -->

## Description

<!-- A few sentences describing the overall goals of the pull request's commits. -->
<!-- If this is a bug fix and you think the fix should be backported, please write so. -->
This PR fixes the issue described in https://github.com/vitessio/vitess/issues/17710.

As the issue describes, the problem happens when the tablet we are trying to promote via `InitPrimary` (can also happen in `PromoteReplica`) times out when writing its tablet record to the topo-server. 
If the topo-server write has succeeded, then the tablet is a primary according to the tablet records, but its internal state doesn't say that its a tablet. Therefore, it keeps publishing to the vtgates that it is a replica. This makes the vtgates think there is no primary tablet and they don't know where to route the queries.

Previously there was no way for VTOrc to detect this situation. It had fixed the mysql level settings by calling `UndoDemotePrimary` but the tablet continued to publish itself as a `Replica` type. 

This PR introduces 2 changes to fix this issue - 
1. The tablets also return their display tablet type in FullStatus output. This information is used by VTOrc to realize that there is a mismatch between the tablet type in the topo record, and the tablet type that the primary thinks it is.
2. `UndoDemotePrimary` also checks that the internal state of the tablet is of type Primary. If not, then it consults the tablet record and if it finds a mismatch then it promotes the tablet without changing the primary term start time.

## Related Issue(s)
- Fixes #17710 
<!-- List related issues and pull requests. If this PR fixes an issue, please add it using Fixes #????  -->

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

## Deployment Notes

<!-- Notes regarding deployment of the contained body of work. These should note any db migrations, etc. -->

timvaillancourt commented:
@GuptaManan100 this is a cool PR 👍 

I wonder if `vttablet` periodically checking-for/fixing the incorrect tablet type itself would be more robust, in addition to the VTOrc recovery. I think VTOrc should be required in a cluster but this would be more resilient in some scenarios where VTOrc is unable to fix this for whatever crazy reason, one edge case I can think of is VTOrc not being present in every cell + network partition, but there are probably many cases

I suppose I'm thinking `vttablet` would [run logic similar to what you added here](https://github.com/vitessio/vitess/pull/17870/files#diff-e00e0cc25918c75778c454c31218b82ae904ebe5fa5060fb39213c550bc17b3cR677-R690) and VTOrc can trigger the same logic. What do you think?

GuptaManan100 commented:
Hey @timvaillancourt 

No I think I would still want VTOrc to make the call. Because changing tablet types are still a cluster event in some sense. Because changing type to primary (with a higher timestamp), is effectively doing a tablet promotion. (Check the test I've added too), and its better to let VTOrc handle it.

Specifically, if we see a newer primary tablet, then VTOrc might choose to run the `NotConnectedToPrimary` recovery instaed of `PrimaryMismatch` that I've added now. This additional context about the shard record, and whether another priary tablet with a higher timestamp exists, is something only VTOrc has, and adding it to vttablet would complicate things and also duplicate the roles.

deepthi commented:
How did this even happen? Did we run out the context clock?
```
			log.Errorf("Error changing type in topo record for tablet %s :- %v\nWill keep trying to read from the toposerver", topoproto.TabletAliasString(ts.tm.tabletAlias), err)
			// In case of a topo error, we aren't sure if the data has been written or not.
			// We must read the data again and verify whether the previous write succeeded or not.
			// The only way to guarantee safety is to keep retrying read until we succeed
			for {
				if ctx.Err() != nil {
					return fmt.Errorf("context canceled updating tablet_type for %s in the topo, please retry", ts.tm.tabletAlias)
				}
				ti, errInReading := ts.tm.TopoServer.GetTablet(ctx, ts.tm.tabletAlias)
				if errInReading != nil {
					<-time.After(100 * time.Millisecond)
					continue
				}
				if ti.Type == tabletType && proto.Equal(ti.PrimaryTermStartTime, PrimaryTermStartTime) {
					log.Infof("Tablet record in toposerver matches, continuing operation")
					break
				}
				log.Errorf("Tablet record read from toposerver does not match what we attempted to write, canceling operation")
				return err
			}
		}
```
Do you know whether a "context canceled ..." error was received?

GuptaManan100 commented:
Yes, we received a context canceled error.

timvaillancourt commented:
> Yes, we received a context canceled error.

I'm fairly sure we've seen the scenario this PR addresses. Extreme CPU load seemed to be the root cause of the inconsistent state 👍
